src/otx/engine/engine_poc.py

```diff
@@ -0,0 +1,67 @@
+import logging
+
+from anomalib.data import AnomalibDataModule
+from anomalib.models import AnomalyModule
+from ultralytics.data import ClassificationDataset as UltralyticsDataset
+from ultralytics.engine.model import Model
+
+from otx.core.data.module import OTXDataModule
+from otx.core.model.base import OTXModel
+from otx.core.utils.cache import TrainerArgumentsCache
+from otx.engine.base import METRICS, Adapter
+
+from .anomalib import AnomalibAdapter
+from .lightning import LightningAdapter
+from .ultralytics import UltralyticsAdapter
+
+logger = logging.getLogger(__name__)
+
+
+class Engine:
+    """Automatically selects the engine based on the model passed to the engine."""
+
+    def __init__(
+        self,
+        **kwargs,
+    ) -> None:
+        self._adapter: Adapter | None = None
+        self._cache = TrainerArgumentsCache(**kwargs)
+
+    @property
+    def adapter(self) -> Adapter:
+        return self._adapter
+
+    @adapter.setter
+    def adapter(self, adapter: adapter) -> None:
+        self._adapter = adapter
+
+    def get_adapter(self, model: OTXModel | AnomalyModule | Model) -> adapter:
+        if isinstance(model, AnomalyModule) and (
+            self.adapter is not isinstance(self.adapter, AnomalibAdapter) or self.adapter is None
+        ):
+            self.adapter = AnomalibAdapter(**self._cache.args)
+        elif isinstance(model, OTXModel) and (self.adapter is None or not isinstance(self.adapter, LightningAdapter)):
+            self.adapter = LightningAdapter(**self._cache.args)
+        elif isinstance(model, Model) and (self.adapter is None or not isinstance(self.adapter, UltralyticsAdapter)):
+            self.adapter = UltralyticsAdapter(**self._cache.args)
+
+        return self.adapter
+
+    def train(
+        self,
+        model: OTXModel | AnomalyModule | Model,
+        datamodule: OTXDataModule | AnomalibDataModule | UltralyticsDataset,
+        **kwargs,
+    ) -> METRICS:
```

Comment:
```python
def train(
        self,
        max_epochs: int = 10,
        seed: int | None = None,
        deterministic: bool | Literal["warn"] = False,
        precision: _PRECISION_INPUT | None = "32",
        callbacks: list[Callback] | Callback | None = None,
        logger: Logger | Iterable[Logger] | bool | None = None,
        resume: bool = False,
        metric: MetricCallable | None = None,
        run_hpo: bool = False,
        hpo_config: HpoConfig = HpoConfig(),  # noqa: B008 https://github.com/omni-us/jsonargparse/issues/423
        checkpoint: PathLike | None = None,
        adaptive_bs: Literal["None", "Safe", "Full"] = "None",
        **kwargs,
    ) -> dict[str, Any]:
```
I like the current way of doing things better, where we clearly define what we need for each command as an argument to each function, as shown above. Also, I don't see any reason to change the model through the function. 

Current OTX: Engine is created (Model is determined) -> E2E flow is used as this model.
This PR's Design: There's room for the Model to change with every command. -> Do we need this use-case?
I believe the above discussion was discussed and decided upon in the first Engine design proposal, and I don't see any reason to change it at this time.

Reply:
Yeah I prefer explicit arguments as well. I didn't include all of them here as I wanted to focus on the adapter. I think we should have explicit arguments also so that these will be serialised by the CLI when creating the YAML file. Maybe this will get polished after a few iterations of discussions.

Reply:
Yes, I see, and the reason I commented was that I thought it would be a good idea to avoid breaking existing Engine Usage as much as possible.